guides/source/caching_with_rails.md

```diff
@@ -538,7 +538,7 @@ config.cache_store = :redis_cache_store, { url: cache_servers,
 
 ### ActiveSupport::Cache::NullStore
 
-This cache store implementation is meant to be used only in development or test environments and it never stores anything. This can be very useful in development when you have code that interacts directly with `Rails.cache` but caching may interfere with being able to see the results of code changes. With this cache store, all `fetch` and `read` operations will result in a miss.
+This cache store is scoped to each web request, and stores values only for the duration of a request. It is meant for use in development and test environments. It can be very useful when you have code that interacts directly with `Rails.cache` but caching interferes with seeing the results of code changes.
```

Comment:
The name "NullStore" is now a bit confusing, since we're documenting that it actually does cache things.

Perhaps this could be closer to the API docs, which mention the LocalStore caching behaviour separately?

https://github.com/rails/rails/blob/2c5c452d8cf3ee8e042aa4c0187b32a88d0da025/activesupport/lib/active_support/cache/null_store.rb#L5-L11

Reply:
Since the local cache strategy isn't introduced in this guide, I'm not sure how I would integrate it here.  I do agree the name `NullStore` isn't particularly fitting, and it may be worth adding a better alias, but I wouldn't want the guidance to be less clear for the sake of the name.

Do you think the following would lessen the dissonance?

```suggestion
This cache store is scoped to each web request, and stored values are cleared at the end of a request. It is meant for use in development and test environments. It can be very useful when you have code that interacts directly with `Rails.cache` but caching interferes with seeing the results of code changes.
```

Reply:
I reworded a bit, putting the focus on "clearing" instead of "storing".  I feel like that reasonably fits the name.  What do you think?

Reply:
Though the text hasn't changed but `caching interferes with seeing the results of code changes` sounds a bit confusing to me.

Reply:
That's marginally better, but I still think this makes the documentation less useful for the average user in order to solve for an artificial scenario (generating random numbers inside a `fetch` block, which won't work in production either).

@topalovic I know it was six months ago, but do you remember the actual problem you had that led to https://github.com/rails/rails/issues/38638?

By "less useful for the average user", I mean that I think people use NullStore when they want to disable caching completely; the current documentation confirms that it's the right store to choose, and it works as long as their `fetch` blocks are deterministic within a single request, which is usually the case. With this change, they'll read "This cache store is scoped to each web request [...]", be forced think about whether that affects them, continue reading the guide in the hope that they'll find a way to _really_ turn off caching, fail to find that option because it doesn't exist...

Reply:
@eugeneius Hey there.

So I was debugging an issue in my view and I did `bin/rails dev:cache` to make sure cache won't interfere with my session. The command replied with:

> Development mode is no longer being cached.

Then it took me half an hour of head scratching to figure out that `Rails.cache.get` is in fact still caching on the request level and that I have no way to bypass that. I perused the docs after the "discovery" and still couldn't understand why `NullStore` would behave this way. Shouldn't it make `Rails.cache.get` a "pass-through"?

The way docs are organized implies (to me at least) that "low-level cache" has nothing to do with web requests, that you can use it in a model or a rake task. So I'd prefer the request persistence documented explicitly, or better yet, have `NullStore` be an actual bit bucket that observes the `dev:cache` toggle. Otherwise, `Null` in `NullStore` is misleading since it is not a bit bucket, but a different kind of "store".

Reply:
Thanks for that explanation. I'm interested in specifics though: what kind of data were you caching such that you expected repeated calls to `Rails.cache.get` with the same key to return different results?

Reply:
It's not really about me expecting repeated calls to `Rails.cache.get` with the same key to return different results. It's about me expecting `bin/rails dev:cache` to disable cache completely.